.github/PULL_REQUEST_TEMPLATE.md

```diff
@@ -0,0 +1,86 @@
+## Type of change
+
+<!-- Mandatory
+Please delete options that are not relevant
+-->
+
+- [ ] New feature
+- [ ] Bug fix
+- [ ] Enhancement
+- [ ] Breaking change
```

Comment:
Because Beats is also a framework we do also distinguish between User changes and Developer changes (if some "public" API is changed).

Reply:
When you say a public API changes, you mean that it could break semver with a breaking change? If that's the case, should we keep `Breaking change` as an item, or should we distinguish between these two: `User changes` and `Developer changes`? I personally like the first one, because in my mind it directly connects with an API change that client code must be aware of.

> I just did give the current template a try: #15422
> 
> For the testing part I wonder if we want to encourage people to use headers for different scenarios/grouping. Do we plan to use the `How to test this PR locally?` section for the test plan?

I'd say that both the review process and the test plan would benefit of having this kind of information. Besides that, explaining how to test something new would make us realise how easy/difficult is to exercise the new code. And not only that, tech writers and end users would read in that section how the software behaves, which is great in terms of gathering expectations.

About multiple test scenarios/groupings, yes, having subheaders (H3, H4...) will definitively help. We can add that to the comments to provide more info

Reply:
> When you say a public API changes, you mean that it could break semver with a breaking change? If that's the case, should we keep Breaking change as an item, or should we distinguish between these two: User changes and Developer changes? I personally like the first one, because in my mind it directly connects with an API change that client code must be aware of.

I wonder what is the motivation using these fields? e.g. could labels also do? Do you plan to process these fields somehow in the future?

We don't really follow semver, because we have to follow common Elastic Stack product versioning. Which means that even minor versions can have breaking changes in the API. We track two changelogs in the Beats project. The actual user facing changelog (Changelog.asciidoc and Changelog.next.asciidoc), and a developer facing one (Changelog-developer.asciidoc and Changelog-developer.next.asciidoc). The developer one tracks changed/fixes/added/breaking.

PRs in Beats sometimes span multiple products. Changes can be developer + user facing at the same time or not at all. That is PRs are not always this well separable by product.

e.g. We can have a 'New feature' in some Beats only while having a 'breaking change' and/or 'new feature' in some API community authors might use in their own Beats.

Reply:
> I wonder what is the motivation using these fields? e.g. could labels also do? Do you plan to process these fields somehow in the future?

Having labels could work too. My preference about having the fields is that is less forgettable than the labels. Anyway, Labels could be used to track issue/pr types.
 
> We don't really follow semver, because we have to follow common Elastic Stack product versioning. Which means that even minor versions can have breaking changes in the API. We track two changelogs in the Beats project. The actual user facing changelog (Changelog.asciidoc and Changelog.next.asciidoc), and a developer facing one (Changelog-developer.asciidoc and Changelog-developer.next.asciidoc). The developer one tracks changed/fixes/added/breaking.
> 
> PRs in Beats sometimes span multiple products. Changes can be developer + user facing at the same time or not at all. That is PRs are not always this well separable by product.
> 
> e.g. We can have a 'New feature' in some Beats only while having a 'breaking change' and/or 'new feature' in some API community authors might use in their own Beats.

I guess this behavior would apply to labels too 🤔 

In any case, we are free to modify/remove any section in this proposal. The goal is to create template which provides great context, fitting Beats unique SDLC. So feel free to suggest adapting the `Type` section in the manner it works better for the team.

Reply:
If we don't plan to use it for automating the release notes I think we should keep it more simple (meaning the way you proposed). The only type I'm missing here is 'Deprecation'.

> Having labels could work too. My preference about having the fields is that is less forgettable than the labels.

+1

Reply:
How about requiring these labels to be present with a bot? I've seen this in other projects and works fairly well. Good message why the PR is not passing + instructions on how to add the right label

Reply:
We could add a comment about how to properly label the PR.

About automating the release notes/changelog, this is a topic we the @elastic/observablt-robots team are working on. We are trying to get a good automated process for that, but not there yet.

Reply:
The bot can add a label based on multiple things. I can imagine:
- if a modified path, then add label Foo (we created this bot!)
- if modified path, then ping Foo user/team
- if not following a structure, then comment the PR
- ...